deltakit-explorer/src/deltakit_explorer/analysis/_analysis.py

```diff
@@ -514,99 +511,196 @@ def calculate_lep_and_lep_stddev(
+        ``scipy.optimize._optimize.OptimizeWarning`` with the message ``"Covariance of
+        the parameters could not be estimated"``. From numerical tests, this only seems
+        to happen when the estimation of Λ satisfies one of the following conditions:
 
+        - ``abs(Λ - 1) < 1e-7``,
+        - Λ is large enough to make the provided ``lep_per_round`` saturate
+          floating-point accuracies.
+
+        Realistically, none of the 2 conditions above can be checked in practice due to
+        sampling noise and sampling overhead, but they might be checked by synthetic
+        data (e.g., in unit-tests).
 
     Examples:
-        Fitting the Lambda value given information for 5, 7, and 9 round of
-        a QEC experiment::
+        Fitting the Λ value given information for 5, 7, and 9 round of a QEC
+        experiment::
 
-            lambda_, lambda_stddev = Analysis.calculate_lambda_and_lambda_stddev(
+            res = calculate_lambda_and_lambda_stddev(
                 distances=[5, 7, 9],
                 lep_per_round=[1.992e-04, 4.314e-05, 7.556e-06],
                 lep_stddev_per_round=[1.2e-05, 9.3e-06, 3.9e-06],
             )
             lambda_, lambda_stddev = res.lambda_, res.lambda_stddev
 
     """
-    query_id = Logging.info_and_generate_uid(locals())
     # Make sure that the inputs are numpy arrays sorted by distance
     isort = np.argsort(distances)
     distances = np.asarray(distances)[isort]
     lep_per_round = np.asarray(lep_per_round)[isort]
     lep_stddev_per_round = np.asarray(lep_stddev_per_round)[isort]
 
-    if len(distances) < 3:
-        error = ValueError("Minimum of 3 distances are required to calculate lambda.")
-        Logging.error(error, query_id)
-        raise error
-    params, cov = np.polyfit(
-        x=distances,
-        y=np.log(lep_per_round),
-        deg=1,
-        w=[lep_pr / lep_std_pr for lep_pr, lep_std_pr in zip(
-            lep_per_round, lep_stddev_per_round)],
-        cov='unscaled'
+    # Check that we do not have duplicate data for the same distance as that will
+    # confuse the numerical methods used in this function.
+    unique_counts = np.unique_counts(distances)
+    non_unique_entries_mask = unique_counts.counts > 1
+    if np.any(non_unique_entries_mask):
+        non_unique_values = unique_counts.values[non_unique_entries_mask].tolist()
+        raise ValueError(
+            "Multiple entries were provided for the following distances: "
+            f"{non_unique_values}. This is not supported. Please make sure you only "
+            "provide one entry per distance."
+        )
+
+    # Make sure that there are no even distances.
+    if np.any(distances % 2 == 0):
+        raise ValueError(
+            "Found at least one even distance in the provided distances "
+            f"({distances.tolist()}). This is not supported."
+        )
+
+    logleppr = np.log(lep_per_round)
+    logleppr_stddev = lep_stddev_per_round / lep_per_round
```

Comment:
Just to confirm, this is also making use of https://en.wikipedia.org/wiki/Propagation_of_uncertainty - if $f = \log{A} \rightarrow \sigma_f = \frac{\sigma_A}{A}$?